File: java/sort/ExternalSort.java
package sort;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Queue;

public class ExternalSort {
    public void sort(String inFile, String outFile) throws IOException {
        int count = readFile(inFile);
        mergeTempFiles(inFile, outFile, count);
        for (int i = 0; i < count; ++i) {
            new File(inFile + "-" + i).delete();
        }
    }

    private int readFile(String inFile) throws IOException {
        List<Integer> values = new ArrayList<>();
        int count = 0;

        try(BufferedReader handle = new BufferedReader(new InputStreamReader(new FileInputStream(inFile)))) {
            String line = "";
            while (null != (line = handle.readLine())) {
                values.add(Integer.parseInt(line));
                if (values.size() >= 65536) {
                    writeTempFiles(values, inFile, count);
                    ++count;
                }
            }
        }

        if (!values.isEmpty()) {
            writeTempFiles(values, inFile, count);
            ++count;
        }
        return count;
    }

    private void mergeTempFiles(String inFile, String outFile, int count) throws IOException {
        Queue<HeapNode> queue = new PriorityQueue<>(cmp);
        for (int i = 0; i < count; ++i) {
            queue.offer(new HeapNode(inFile, i));
        }

        try(BufferedWriter handle = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(outFile, false)))) {
            while (!queue.isEmpty()) {
                HeapNode heap = queue.poll();
                handle.write(heap.value + "\n");

                String line = heap.handle.readLine();
                if (null == line) {
                    heap.handle.close();
                } else {
                    heap.value = Integer.parseInt(line);
                    queue.offer(heap);
                }
            }
        }
    }

    private void writeTempFiles(List<Integer> values, String inFile, int index) throws IOException {
        Collections.sort(values);

        try(BufferedWriter handle = new BufferedWriter(new OutputStreamWriter(new FileOutputStream(inFile + "-" + index, false)))) {
            for (int value : values) {
                handle.write(value + "\n");
            }
        }
        values.clear();
    }

    private class HeapNode {
        public int value = 0;
        public BufferedReader handle = null;

        public HeapNode(String inFile, int index) throws IOException {
            handle = new BufferedReader(new InputStreamReader(new FileInputStream(inFile + "-" + index)));
            value = Integer.parseInt(handle.readLine());
        }
    }

    private static Comparator<HeapNode> cmp = new Comparator<HeapNode>() {
        @Override
        public int compare(HeapNode node1, HeapNode node2) {
            return node1.value - node2.value;
        }
    };
}

File: py/sort/ExternalSort.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import heapq, os
from typing import List

class ExternalSort:
    def sort(self, inFile: str, outFile: str) -> None:
        count = self.__readFile(inFile)
        self.__mergeTempFiles(inFile, outFile, count)
        for i in range(count):
            os.remove(inFile + '-' + str(i))

    def __readFile(self, inFile: str) -> int:
        values, count = [], 0
        with open(inFile, 'r') as handle:
            for line in handle:
                values.append(int(line))
                if len(values) >= 65536:
                    self.__writeTempFiles(values, inFile, count)
                    count += 1

        if values:
            self.__writeTempFiles(values, inFile, count)
            count += 1
        return count

    def __mergeTempFiles(self, inFile: str, outFile: str, count: int) ->None:
        queue = []
        for i in range(count):
            heapq.heappush(queue, self.__HeapNode(inFile, i))

        with open(outFile, 'w') as handle:
            while queue:
                heap = heapq.heappop(queue)
                handle.write(str(heap.value) + '\n')

                line = heap.handle.readline()
                if not line:
                    heap.handle.close()
                else:
                    heap.value = int(line)
                    heapq.heappush(queue, heap)

    def __writeTempFiles(self, values: List[int], inFile: str, index: int) -> None:
        values.sort()

        with open(inFile + '-' + str(index), 'w') as handle:
            for value in values:
                handle.write(str(value) + '\n')
        values.clear()

    class __HeapNode:
        def __init__(self, inFile: str, index: int):
            self.handle = open(inFile + '-' + str(index), 'r')
            self.value = int(self.handle.readline())

        def __lt__(self, other: '__HeapNode'):
            return self.value < other.value
